OrbitVulkanLayer/VulkanLayerController.h

```diff
@@ -0,0 +1,460 @@
+    if (layer_name != nullptr && strcmp(layer_name, kLayerName) == 0) {
+      if (property_count != nullptr) {
+        *property_count = 0;
+      }
+      return VK_SUCCESS;
+    }
+
+    // Vulkan spec mandates returning this when this layer isn't being queried.
+    return VK_ERROR_LAYER_NOT_PRESENT;
+  }
+
+  [[nodiscard]] VkResult OnEnumerateDeviceExtensionProperties(VkPhysicalDevice physical_device,
+                                                              const char* layer_name,
+                                                              uint32_t* property_count,
+                                                              VkExtensionProperties* properties) {
+    // If our layer is queried exclusively, we just return our extensions. Note that queries with
+    // layer_name == nullptr request all extensions.
+    if (layer_name != nullptr && strcmp(layer_name, kLayerName) == 0) {
+      // If properties == nullptr, only the number of extensions is queried.
+      if (properties == nullptr) {
+        *property_count = kDeviceExtensions.size();
+        return VK_SUCCESS;
+      }
+      uint32_t num_extensions_to_copy = kDeviceExtensions.size();
+      // In the case that less extensions are queried then the layer uses, we copy on this number
+      // and return VK_INCOMPLETE, according to the specification.
+      if (*property_count < num_extensions_to_copy) {
+        num_extensions_to_copy = *property_count;
+      }
+      memcpy(properties, kDeviceExtensions.data(),
+             num_extensions_to_copy * sizeof(VkExtensionProperties));
+      *property_count = num_extensions_to_copy;
+
+      if (num_extensions_to_copy < kDeviceExtensions.size()) {
+        return VK_INCOMPLETE;
+      }
+      return VK_SUCCESS;
+    }
+
+    // If a different layer is queried exclusively, we forward the call.
+    if (layer_name != nullptr) {
+      return dispatch_table_.EnumerateDeviceExtensionProperties(physical_device)(
+          physical_device, layer_name, property_count, properties);
+    }
+
+    // This is a general query, so we need to append our extensions to the ones down in the
+    // callchain.
+    uint32_t num_other_extensions;
+    VkResult result = dispatch_table_.EnumerateDeviceExtensionProperties(physical_device)(
+        physical_device, nullptr, &num_other_extensions, nullptr);
+    if (result != VK_SUCCESS) {
+      return result;
+    }
+
+    std::vector<VkExtensionProperties> extensions(num_other_extensions);
+    result = dispatch_table_.EnumerateDeviceExtensionProperties(physical_device)(
+        physical_device, nullptr, &num_other_extensions, extensions.data());
+    if (result != VK_SUCCESS) {
+      return result;
+    }
+
+    // Lets append all of our extensions, that are not yet listed.
+    // Note as this list of our extensions is very small, we are fine with O(N*M) runtime.
+    for (const auto& extension : kDeviceExtensions) {
+      bool is_unique = true;
```

Comment:
Nit: I find this a bit unclear, but also maybe I'm misunderstanding. If this were:

bool already_present = false;
for () {
   if (strcmp(...)) {
     already_present = true;
   }
} 

if (!already_present) {
...
}

I think it'd a bit easier to understand.

Reply:
Done.